cluster-autoscaler/core/resource_slack.go

```diff
@@ -0,0 +1,227 @@
+	toGB := func(value float64) float64 {
+		return value / 1024 / 1024 / 1024
+	}
+
+	// TODO Could be simplified more
+	// Ty our best to make all the words match to the original design at https://github.com/kubernetes/contrib/issues/2251#issuecomment-272893268
+	clusterSize := int64(len(readyNodes))
+	clusterSizeInFloat := float64(clusterSize)
+	// Foundamental statistics used for creating placeholder pods matching the reality
+	// Use Quantity where possible not to introduce bugs due to numeric conversion among different formats
+	clusterCpuSum := &resource.Quantity{}
+	clusterMemorySum := &resource.Quantity{}
+	minNodeMilliCPU := int64(math.MaxInt64)
+	minNodeMemory := int64(math.MaxInt64)
+	for _, n := range readyNodes {
+		if cpu := n.Status.Capacity.Cpu(); cpu != nil {
+			clusterCpuSum.Add(*cpu)
+			v := cpu.MilliValue()
+			if v < minNodeMilliCPU {
+				minNodeMilliCPU = v
+			}
+		}
+		if mem := n.Status.Capacity.Memory(); mem != nil {
+			clusterMemorySum.Add(*mem)
+			v := mem.Value()
+			if v < minNodeMemory {
+				minNodeMemory = v
+			}
+		}
+	}
+	// Resource occupied by each node
+	avgNodeMilliCPU := clusterCpuSum.MilliValue() / clusterSize
+	avgNodeMemory := float64(clusterMemorySum.Value()) / clusterSizeInFloat
+	// Resource occupied by the cluster
+	clusterMilliCPU := avgNodeMilliCPU * clusterSize
+	clusterMemory := avgNodeMemory * clusterSizeInFloat
+	// Resource should be occupied by placeholder pods
+	minExtraCapacityRate := autoscalingContext.MinExtraCapacityRate
+	clusterExtraMilliCPU := int64(float64(clusterMilliCPU) * minExtraCapacityRate)
+	clusterExtraMemory := clusterMemory * minExtraCapacityRate
+	// Mix different types of placeholder pods to ensure both:
+	// (1) providing extra capacity even for the largest pod in the cluster and
+	// (2) not wasting cluster capacity by making all the placeholder pods too big
+	//
+	// * Note that (1) could be made configurable if necessary. How about capping the max size of a placeholder pod via
+	// flags --placeholder-pod-max-cpu and --placeholder-pod-max-memory so that CA won't waste money by keeping
+	// considerably large nodes for the latest pod.
+	//
+	// In a nutshell, we prefer choosing a node group with a largest machine type for scaling up so that
+	// every pod can have a chance to be scheduled faster. Choosing a too small machine type for scaling up result in
+	// large pods to not fit within extra capacity, which breaks the original purpose of the resource slack feature(--min-extra-capacity-rate)
+	//
+	// General rules
+	// * Every place holder pod should be small enough to fit within at least one of nodes
+	//   For example, a 30GB placeholder pod which doesn't fit to any small node which results in more nodes doesn't make sense.
+	// * Exactly one place holder pod should be sized the same as the largest(in terms of resource requests) pod
```

Comment:
This is tricky. Imagine that you have a stateful set pod that requires 10cpu and 1gb of memory and 1 cpu and 100 gb of memory, both placed on machines that are not autoscaled. Your code will try to create a pod with 10 cpus and 100 gb and fit it on the autoscaled machines. 
For the first iteration I would rather NOT have this type of heuristic as it might be hard to do right.